src/Zendesk/Functions.hs

```diff
@@ -0,0 +1,174 @@
+import           Network.HTTP.Simple (Request, addRequestHeader, getResponseBody, httpJSON, httpLBS,
+                                      parseRequest_, setRequestBasicAuth, setRequestBodyJSON,
+                                      setRequestMethod, setRequestPath)
+
+import           Zendesk.Types (Attachment (..), Comment (..), RequestType (..), Ticket (..),
+                                TicketId, TicketInfo (..), TicketList (..), TicketTag (..),
+                                ZendeskLayer (..), parseAgentId, parseComments, parseTickets,
+                                renderTicketStatus, Config (..))
+
+
+-- | The default configuration.
+defaultConfig :: Config
+defaultConfig =
+    Config
+        { cfgAgentId            = 0
+        , cfgZendesk            = "https://iohk.zendesk.com"
+        , cfgToken              = ""
+        , cfgEmail              = "daedalus-bug-reports@iohk.io"
+        , cfgAssignTo           = 0
+        , cfgKnowledgebase      = []
+        , cfgNumOfLogsToAnalyze = 5
+        , cfgIsCommentPublic    = True -- TODO(ks): For now, we need this in CLI.
+        , cfgZendeskLayer       = basicZendeskLayer
+        }
+
+-- | The basic Zendesk layer.
+basicZendeskLayer :: (MonadIO m, MonadReader Config m) => ZendeskLayer m
+basicZendeskLayer = ZendeskLayer
+    { zlGetTicketInfo           = getTicketInfo
+    , zlListTickets             = listTickets
+    , zlPostTicketComment       = postTicketComment
+    , zlGetAgentId              = getAgentId
+    , zlGetAttachment           = getAttachment
+    , zlGetTicketComments       = getTicketComments
+    }
+
+
+-- | Get single ticket info.
+getTicketInfo
+    :: (MonadIO m, MonadReader Config m)
+    => TicketId
+    -> m TicketInfo
+getTicketInfo ticketId = do
+    cfg <- ask
+
+    let req = apiRequest cfg ("tickets/" <> show ticketId <> ".json")
+    liftIO $ apiCall parseJSON req
+
+-- | Return list of ticketIds that has been requested by config user (not used)
+listTickets
+    :: (MonadIO m, MonadReader Config m)
+    => RequestType
+    -> m [TicketInfo]
+listTickets request = do
+    cfg <- ask
+
+    let agentId = cfgAgentId cfg
+    let url = case request of
+                  Requested -> "/users/" <> show agentId <> "/tickets/requested.json"
+                  Assigned  -> "/users/" <> show agentId <> "/tickets/assigned.json"
+    let req = apiRequest cfg url
+
+    let go :: [TicketInfo] -> Text -> IO [TicketInfo]
+        go list' nextPage' = do
+          let req' = apiRequestAbsolute cfg nextPage'
+          (TicketList pagen nextPagen) <- apiCall parseTickets req'
+          case nextPagen of
+              Just nextUrl -> go (list' <> pagen) nextUrl
+              Nothing      -> pure (list' <> pagen)
+
+    (TicketList page0 nextPage) <- liftIO $ apiCall parseTickets req
+    case nextPage of
+        Just nextUrl -> liftIO $ go page0 nextUrl
+        Nothing      -> pure page0
+
+-- | Send API request to post comment
+postTicketComment
```

Comment:
Since we've defined `ZendeskResponse` in the previous PR, we can make this function to take `ZendeskResponse` as an argument which will be much more simpler
`postTicketComment
    :: (MonadIO m, MonadReader Config m)
    => ZendeskResponse
    -> m ()`

Reply:
Nice, thanks.